### PR TITLE
Make boa::parse emit error on invalid input, not panic

### LINKED ISSUE #772: panic on invalid input to boa::parse, 'Could not convert to BigInt'
**Describe the bug**

Expected an Error, not a panic.

**To Reproduce**

Can be reproduced with this program

```rust
fn main() {
    let data: Vec<u8> = vec![0x30, 0x78, 0x6e, 0x3d, 0x3e, 0x78, 0x78]; // "0xn=>xx"
    if let Ok(s) = std::str::from_utf8(&data) {
        let _ = boa::parse(s);
    }
}
```

**Expected behavior**

An Error, not a panic

**Build environment (please complete the following information):**

- OS: Ubuntu 20.04
- Version: 0.10.0
- Target triple: [e.g. x86_64-unknown-linux-gnu]
- Rustc version: 1.48.0-nightly (d006f5734 2020-08-28)

**Additional context**
Full stacktrace:
```
thread 'main' panicked at 'Could not convert to BigInt', /home/capitol/.cargo/registry/src/github.com-1ecc6299db9ec823/Boa-0.10.0/src/syntax/lexer/number.rs:291:59
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d006f5734f49625c34d6fc33bf6b9967243abca8/library/std/src/panicking.rs:483
   1: core::panicking::panic_fmt
             at /rustc/d006f5734f49625c34d6fc33bf6b9967243abca8/library/core/src/panicking.rs:85
   2: core::option::expect_failed
             at /rustc/d006f5734f49625c34d6fc33bf6b9967243abca8/library/core/src/option.rs:1213
   3: core::option::Option<T>::expect
             at /home/capitol/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:333
   4: <boa::syntax::lexer::number::NumberLiteral as boa::syntax::lexer::Tokenizer<R>>::lex
             at /home/capitol/.cargo/registry/src/github.com-1ecc6299db9ec823/Boa-0.10.0/src/syntax/lexer/number.rs:291
   5: boa::syntax::lexer::Lexer<R>::next
             at /home/capitol/.cargo/registry/src/github.com-1ecc6299db9ec823/Boa-0.10.0/src/syntax/lexer/mod.rs:194
   6: boa::syntax::parser::cursor::buffered_lexer::BufferedLexer<R>::fill
             at /home/capitol/.cargo/registry/src/github.com-1ecc6299db9ec823/Boa-0.10.0/src/syntax/parser/cursor/buffered_lexer/mod.rs:116
   7: boa::syntax::parser::cursor::buffered_lexer::BufferedLexer<R>::peek
             at /home/capitol/.cargo/registry/src/github.com-1ecc6299db9ec823/Boa-0.10.0/src/syntax/parser/cursor/buffered_lexer/mod.rs:201
   8: boa::syntax::parser::cursor::Cursor<R>::peek
             at /home/capitol/.cargo/registry/src/github.com-1ecc6299db9ec823/Boa-0.10.0/src/syntax/parser/cursor/mod.rs:56
   9: <boa::syntax::parser::Script as boa::syntax::parser::TokenParser<R>>::parse
             at /home/capitol/.cargo/registry/src/github.com-1ecc6299db9ec823/Boa-0.10.0/src/syntax/parser/mod.rs:124
  10: boa::syntax::parser::Parser<R>::parse_all
             at /home/capitol/.cargo/registry/src/github.com-1ecc6299db9ec823/Boa-0.10.0/src/syntax/parser/mod.rs:104
  11: boa::parse
             at /home/capitol/.cargo/registry/src/github.com-1ecc6299db9ec823/Boa-0.10.0/src/lib.rs:73
  12: boa_reproduce::main
             at ./src/main.rs:4
  13: core::ops::function::FnOnce::call_once
             at /home/capitol/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
```

### PR DESCRIPTION
**Pull request by gorogoroumaru**
This Pull Request fixes/closes #772 .

It changes the following:

- boa::parse panics on invalid input. This behavior is not expected, and I fixed this to emit error on invalid input
- fixed typo in boa/src/syntax/ast/op.rs

### CODE REVIEW

File: boa/src/syntax/lexer/number.rs [hunk 1]
```diff
@@ -287,42 +287,47 @@ impl<R> Tokenizer<R> for NumberLiteral {
 
         let num = match kind {
             NumericKind::BigInt(base) => {
-                Numeric::BigInt(
-                    BigInt::from_string_radix(&buf, base).expect("Could not convert to BigInt")
-                    )
+                let val = BigInt::from_string_radix(&buf, base);
+                match val {
+                    Some(bigint) => Ok(Numeric::BigInt(bigint)),
+                    _ => Err("Could not convert to BigInt"),
```
Comment (HalidOdat) [inline boa/src/syntax/lexer/number.rs:293]: We should not be changing this to return `Err` if it reached this point and it gave an error from `from_string_radix` that means there is a bug it the logic on top.

Here is the problem with the input `0xn` is it sets it as hexadecimal number kind, so we reach the
```
  ⬇️ 
0xn
```
and we reach the `n` we and in line `230` we consume all the number characters, and the problem is here
we need to check that at least there is one digit that has been consumed. when we are in non-decaimal mode (in base 16, 8, 2).

This error is generated also with `0on` (octal) and `0bn` (binary)

Comment (gorogoroumaru) [reply boa/src/syntax/lexer/number.rs:293]: Thank you @HalidOdat!
Could you guide me how to solve this?

Comment (HalidOdat) [reply boa/src/syntax/lexer/number.rs:293]: > Thank you @HalidOdat!
> Could you guide me how to solve this?

We need to peek at the char before line `230` where we start consuming the digit chars: we need to check if we are in not decimal mode, if so we peek the next char if it is not a digit according to the base that its in (16, 8, 2) and if that is the case we generate an `Err`, if not we just fall through.

Comment (HalidOdat) [reply boa/src/syntax/lexer/number.rs:293]: if im not mistaken the cursor has a `.peek()` method, and we can get the base with `kind.base()`, and check if its a digit with that base with `char::is_digit(base)`

File: boa/src/syntax/lexer/number.rs [hunk 2]
```diff
@@ -226,6 +226,17 @@ impl<R> Tokenizer<R> for NumberLiteral {
...
 
+        //Checks if the next
+        let digit_char = cursor.peek();
+
+        if let Some(digit) = digit_char? {
+            if !digit.is_digit(kind.base()) {
+                return Err(Error::syntax(
+                    "character was not in expected digit",
+                    cursor.pos(),
+                ));
```
Comment (HalidOdat) [inline boa/src/syntax/lexer/number.rs:237]:
```suggestion
                return Err(Error::syntax(
                    "expected digit after number base prefix",
                    cursor.pos(),
                ));
```

File: boa/src/syntax/lexer/number.rs [hunk 3]
```diff
@@ -226,6 +226,17 @@ impl<R> Tokenizer<R> for NumberLiteral {
             }
         }
 
+        //Checks if the next
+        let digit_char = cursor.peek();
+
+        if let Some(digit) = digit_char? {
```
Comment (HalidOdat) [inline boa/src/syntax/lexer/number.rs:232]:
```suggestion
        // Checks if the next char after '0x', '0o' or '0b' is a digit
        // of that base. if not return an error. 
        if let Some(digit) = cursor.peek()? {
```

File: boa/src/syntax/lexer/number.rs [hunk 4]
```diff
@@ -226,6 +226,16 @@ impl<R> Tokenizer<R> for NumberLiteral {
...
+        // Checks if the next char after '0x', '0o' or '0b' is a digit
+        // of that base. if not return an error. 
+        if let Some(digit) = cursor.peek()? {
+            if !digit.is_digit(kind.base()) {
+                return Err(Error::syntax(
+                    "expected digit after number base prefix",
+                    cursor.pos(),
+                ));
+            }
+        }
```
Comment (Lan2u) [inline boa/src/syntax/lexer/number.rs:238]: I think this is missing checks that the base is 2/8/16 and also this should be moved up into the if (0) statement above

Comment (HalidOdat) [reply boa/src/syntax/lexer/number.rs:238]: > I think this is missing checks that the base is 2/8/16 and also this should be moved up into the if (0) statement above,

Probably best if we move this to the `0x`, `0o`, and `0b` cases this should also remove some check, what base it should be and and what base the digit should be.

so on line 158 (hexadecimal 0x) we peek a char and if it is not a digit with base `16` we return error "expected hexadecimal digit after '0x' prefix"

on line 166 (ocatal 0o) we peek a char and if it is not a digit with base `8` we return error "expected octal digit after '0o' prefix"

on line 174 (binary 0b) we peek a char and if it is not a digit with base `2` we return error "expected binary digit after '0b' prefix"

This approach has the added benefit of having clearer error messages

**Comment by gorogoroumaru** [comment]
Thank you @Lan2u and @HalidOdat!
I applied your suggested changes.

**Review by HalidOdat — APPROVED** [review]

**Review by Lan2u — APPROVED** [review]

**Pull request merged**